src/python_testing/matter_testing_support.py

```diff
@@ -436,55 +451,52 @@ def as_string(self, mapper: ClusterMapper):
         return desc
 
     def __str__(self):
-        return (f'\n        Endpoint: {self.endpoint_id},'
-                f'\n        Cluster:  {cluster_id_str(self.cluster_id)},'
-                f'\n        Attribute:{id_str(self.attribute_id)}')
+        return (f'{super().__str__()}'
+                f'\n      Attribute:{id_str(self.attribute_id)}')
 
 
 @dataclass
-class EventPathLocation:
-    endpoint_id: int
-    cluster_id: int
+class EventPathLocation(ClusterPathLocation):
     event_id: int
 
     def __str__(self):
-        return (f'\n        Endpoint: {self.endpoint_id},'
-                f'\n        Cluster:  {cluster_id_str(self.cluster_id)},'
-                f'\n        Event:    {id_str(self.event_id)}')
+        return (f'{super().__str__()}'
+                f'\n       Event:    {id_str(self.event_id)}')
 
 
 @dataclass
-class CommandPathLocation:
-    endpoint_id: int
-    cluster_id: int
+class CommandPathLocation(ClusterPathLocation):
     command_id: int
 
     def __str__(self):
-        return (f'\n        Endpoint: {self.endpoint_id},'
-                f'\n        Cluster:  {cluster_id_str(self.cluster_id)},'
-                f'\n        Command:  {id_str(self.command_id)}')
+        return (f'{super().__str__()}'
+                f'\n       Command:  {id_str(self.command_id)}')
 
 
 @dataclass
-class ClusterPathLocation:
-    endpoint_id: int
-    cluster_id: int
+class FeaturePathLocation(ClusterPathLocation):
+    feature_code: str
 
     def __str__(self):
-        return (f'\n       Endpoint: {self.endpoint_id},'
-                f'\n       Cluster:  {cluster_id_str(self.cluster_id)}')
+        return (f'{super().__str__()}'
+                f'\n       Feature:  {self.feature_code}')
 
 
 @dataclass
-class FeaturePathLocation:
-    endpoint_id: int
-    cluster_id: int
-    feature_code: str
+class DeviceTypePathLocation(ProblemLocation):
+    device_type_id: int
+    cluster_id: Optional[int]
+
+    def __init__(self, device_type_id: int, cluster_id: Optional[int] = None):
```

Comment:
init should not be needed for dataclasses. I guess this may be due to using a base class? I am suggesting not using a base class in this instance.

Reply:
You can set the default for cluster_id in the type `i.e. cluster_id: Optional[int] = None` and then this constructor should be auto-created for you.

Reply:
fixed.